docs/src/formula.md

```diff
@@ -14,167 +14,184 @@ fields with possibly heterogeneous types.  One of the primary goals of
+`fit` methods that take a `FormulaTerm` and a table instead of numeric
+matrices.
 
-An equivalent but slightly more verbose way of doing the same thing is:
+As an example, we'll simulate some data from a linear regression model with an
+interaction term, a continuous predictor, a categorical predictor, and the
+interaction of the two, and then fit a `GLM.LinearModel` to recover the
+simulated coefficients.
 
-```jldoctest
-julia> formula_ex = :(@formula($lhs ~ 1 + x))
-:(@formula y ~ 1 + x)
+```julia-repl
+julia> using GLM, DataFrames, StatsModels
 
-julia> eval(formula_ex)
-Formula: y ~ 1 + x
-```
+julia> data = DataFrame(a = rand(100), b = repeat(["d", "e", "f", "g"], 25));
 
-### Technical details
+# simulate response:
+julia> X = modelmatrix(@formula(y ~ 1 + a*b).rhs, data);
 
-You may be wondering why formulas in Julia require a macro, while in R they
-appear "bare."  R supports nonstandard evaluation, allowing the formula to
-remain an unevaluated object while its terms are parsed out. Julia uses a much
-more standard evaluation mechanism, making this impossible using normal
-expressions. However, unlike R, Julia provides macros to explicitly indicate
-when code itself will be manipulated before it's evaluated. By constructing a
-formula using a macro, we're able to provide convenient, R-like syntax and
-semantics.
+julia> β_true = [1., 2, 3, 4, 5, 6, 7, 8];
 
-The formula syntactic transformations are applied _at parse time_ when using the
-`@formula` macro.  You can see this with using `@macroexpand`:
+julia> ϵ = randn(100)*0.1;
 
-```jldoctest
-@macroexpand @formula y ~ 1 + (a+b)*c
-:((StatsModels.Formula)($(Expr(:copyast, :($(QuoteNode(:(y ~ 1 + (a + b) * c)))))), $(Expr(:copyast, :($(QuoteNode(:(y ~ 1 + a + b + c + a & c + b & c)))))), :y, $(Expr(:copyast, :($(QuoteNode(:(1 + a + b + c + a & c + b & c))))))))
-```
-Or more legibly
-```julia
-:((StatsModels.Formula)($(Expr(:copyast, :($(QuoteNode(:(y ~ 1 + (a + b) * c)))))),
-                        $(Expr(:copyast, :($(QuoteNode(:(y ~ 1 + a + b + c + a & c + b & c)))))),
-                        :y,
-                        $(Expr(:copyast, :($(QuoteNode(:(1 + a + b + c + a & c + b & c))))))))
-```
+julia> data[:y] = X*β_true .+ ϵ
 
-The `@formula` macro re-writes the formula expression `y ~ 1 + (a+b)*c` as a
-call to the `Formula` constructor.  The arguments of the constructor correspond
-to the fields of the `Formula` struct, which are, in order:
+julia> fit(LinearModel, @formula(y ~ 1 + a*b), data)
+StatsModels.TableRegressionModel{LinearModel{LmResp{Array{Float64,1}},DensePredChol{Float64,LinearAlgebra.Cholesky{Float64,Array{Float64,2}}}},Array{Float64,2}}
 
-* `ex_orig`: the original expression `:(y ~ 1+(a+b)*c)`
-* `ex`: the parsed expression `:(y ~ 1+a+b+a&c+b&c)`
-* `lhs`: the left-hand side `:y`
-* `rhs`: the right-hand side `:(1+a+b+a&c+b&c)`
+y (continuous) ~ 1 + a (continuous) + b (4 levels): DummyCoding(3) + a (continuous) & b (4 levels): DummyCoding(3)
```

Comment:
Isn't this too verbose? Even for a simple model like here, the line doesn't fit on a standard screen. Also in this context, we can just check the coefficient table below to see how each variable has been expanded.

Reply:
yeah I've considered that.  one option would be to add a check for a compact io context (and use that when showing a formula term or when showing a model) and just print something abbreviated.

Reply:
(and by "considered" I mean also worried about it, but haven't acted in those worries yet...)

Reply:
Good idea.

Though in addition to that the verbose printing could be compressed a bit. Continuous variables could be treated as the default, so that they don't need a specific mention (and indeed they correspond to a single column in the matrix). And for categorical variables, the number of levels doesn't have to be repeated in addition to the coding (even though it can be kept when printing the term in isolation). Finally, to make it easier to parse the formula, it would make sense to have spaces only around operators, but not between a term and its description.